### PR TITLE
Phase 1 of automatic content moderation - lets get some data

### PR DESCRIPTION
**Pull request by dylanjha**
This will attach a **moderation score** to the slack message with the following parameters:

* adult
* violent
* racy

Scores range from 1-5.

moderation-score_2021-05-20_15-08-55

The way this works:

* get 3 sample thumbnails: at the 1/4th mark, halfway mark and 3/4th mark of the video
* run each thumbnail through google cloud vision and convert the score 1-5 based on likelihood
* take the max of each dimension

After merging, the plan would be to:

* Run this in production, monitor the slackbot moderation channel and see how the results are and where we have to draw the line
* After we know where to draw the line, we can continue the work that was started on [this branch](https://github.com/muxinc/stream.new/compare/main...dylanjha:dj/automatic-moderate?expand=1) to automatically disable playback IDs that get flagged above a certain threshold

TODO:

- [ ] load google credentials via an env var, not through the file system
- [ ] set this up in the right DevEx google account, let's not get rate limited or something

### CODE REVIEW

File: lib/google-vision-client.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+process.env.GOOGLE_APPLICATION_CREDENTIALS = process.env.GOOGLE_APPLICATION_CREDENTIALS || 'three-monkeys-hackweek-844be021a66d.json';
+import vision from '@google-cloud/vision';
+
+const client = new vision.ImageAnnotatorClient();
```
Comment (dylanjha) [inline lib/google-vision-client.ts:4]: I was unable to use [nock](https://github.com/nock/nock) in tests. It just straight up was unaware of the HTTP requests that were made by google's library :(

I think b/c google under the hood is not using node's `http` module.

Anyone know more about this than I do?

**Review by jsanford8 — APPROVED** [review]

File: lib/moderation.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  };
+
+  return combined;
+}
+
+export async function getScores ({ playbackId, duration }: { playbackId: string, duration: number }): Promise<ModerationScores> {
+  /*
+   * Get 3 thumbnails based on the duration
+   */
+  const timestamps = [(duration * 0.25),  (duration * 0.5), (duration * 0.75)];
```
Comment (mmcc) [inline lib/moderation.ts:88]: This is a small nit, but I think we should consider randomizing the exact timestamps. Doesn't have to be on this first pass, but longer-term I think we'll want to randomize the selection within those 3 segments of the video.

**Review by mmcc — COMMENTED** [review]
👍 ship it.

File: lib/moderation.ts (hunk 2 above)
Comment (dylanjha) [reply lib/moderation.ts:88]: That's a good idea! Something like this? Trying to come up with an elegant way but also need some safeguards to make sure it doesn't break or end up with negative numbers on an asset with a super short duration

```js
// returns a random integer between min and max
function generateRandomInteger (min, max) {
  return Math.floor(min + Math.random()*(max + 1 - min));
}

// randomly make a number either positive or negative
function positiveOrNegative (num) {
  return ((Math.random() - 0.5) * 2 > 0) ? num : (num * -1)
}

const timestamps = [(duration * 0.25),  (duration * 0.5), (duration * 0.75)].map((timestamp) => {
  // jitter will be something between 1/10 and 1/20 of the duration of the video
  // a 5 minute video is 300 seconds, jitter will be between 15 sec (1/20th) and 30 sec (1/10th)
  // we'll take that jitter and make it positive or negative and shift the timestamp value
  const jitter = duration / generateRandomInteger(10, 20);
  return timestamp + positiveOrNegative(jitter);
})
```

**Pull request merged**